include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +502,458 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+{
+	return irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotH * NdotH, TdotH * TdotH, BdotH * BdotH, TdotL * TdotL, BdotL * BdotL, TdotV * TdotV, BdotV * BdotV, NdotV * NdotV, NdotL * NdotL, nx, ny, ax2, ay2);
+}
+float irr_glsl_blinn_phong_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache microfacet, in irr_glsl_AnisotropicViewSurfaceInteraction i, in float nx, in float ny, in float ax2, in float ay2)
+{
+	return irr_glsl_blinn_phong_cos_eval_DG(
+		microfacet.isotropic.NdotH,
+		microfacet.TdotH,
+		microfacet.BdotH,
+		s.NdotL,
+		s.TdotL,
+		s.BdotL,
+		i.isotropic.NdotV,
+		i.TdotV,
+		i.BdotV,
+		nx, ny, ax2, ay2
+	);
+}
+
+#ifdef GEN_CHOICE_STREAM
+void instr_eval_and_pdf_execute(in instr_t instr, in MC_precomputed_t precomp, in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache _microfacet)
+{
+	const uint op = instr_getOpcode(instr);
+	const bool is_bxdf = op_isBXDF(op);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons) [(is_bxdf && is_bsdf) actually tells that op is bsdf]
+	// (is_bxdf && is_bsdf) -> BSDF
+	// (is_bxdf && !is_bsdf) -> BRDF
+	const bool is_bxdf_or_combiner = op_isBXDForCoatOrBlend(op);
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+
+	uvec3 regs = instr_decodeRegisters(instr);
+	params_t params;
+	bsdf_data_t bsdf_data;
+	mat2x3 ior;
+	mat2x3 ior2;
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))//does cos>0 check even has any point here? L comes from a sample..
+	if (is_bxdf_or_combiner)
 	{
-#ifdef OP_DIFFUSE
-		if (op==OP_DIFFUSE) {
-			pdf = irr_glsl_oren_nayar_pdf(s, currInteraction.isotropic, a2);
-		} else
-#endif
+		bsdf_data = fetchBSDFDataForInstr(instr);
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
+		microfacet = _microfacet;
+	}
 
-#ifdef NDF_GGX
-		if (ndf==NDF_GGX) {
+	vec3 eval = vec3(0.0);
+	float pdf = 0.0;
 
+	if (is_bxdf)
+	{
+		//speculative execution
+		uint ndf = instr_getNDF(instr);
+		float a = params_getAlpha(params);
+		float a2 = a*a;
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
-			pdf = irr_glsl_ggx_pdf(s, currInteraction.isotropic, a2);
+		float one_minus_a2 = 1.0 - a2;
 #else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-			pdf = irr_glsl_ggx_pdf(s, currInteraction, a, ay, a2, ay2);
+		float ay = params_getAlphaV(params);
+		float ay2 = ay*ay;
 #endif
 
-		} else
+		const float eta = colorToScalar(ior[0]);
+		const float rcp_eta = 1.0 / eta;
+		const vec3 albedo = params_getReflectance(params);
```

Comment:
move albedo fetching into the `if (is_valid)`

also start off with `eta` and `rcp_eta` uninitialized.. only initialize them inside `if (is_bsdf & isTransmissionPath)¬

Reply:
it's not fetching per se, fetching is `instr_getParameters()`. Seems like parameter fetching is much less divergent than you think
Second thing: im confused, first you say to put `op_isDiffuse(op)` outside is_valid check, but you want albedo inside is_valid check

Reply:
> Second thing: im confused, first you say to put op_isDiffuse(op) outside is_valid check, but you want albedo inside is_valid check

My bad.. `is_valid` is only for microfacet models, diffuse only need to check that clamped cosine is `>FLT_MIN`

Then I guess albedo needs to stay outside as you've pointed out

Reply:
ok so its good